### ISSUE TITLE
[ORD-5.1.1.1] Reaction Preview/Summary

### ISSUE THREAD

**Issue by yauhen-karmyzau** [open-reaction-database/ord-app#53]
Tasks for analysis:
- [x] #54
- [x] #55
- [x] #161

# User Story # ORD-5.1.1.1 & 6.1.1

## 2. Revision History

| Version | Date       | Description |
| ------- | ---------- | ----------- |
| V1.0    | 20.12.2024 | Discovery Phase: The user should be able to preview the reaction, with the option to copy the reaction as a .png image. The reaction preview should be visually organized, clearly separating components and products, allowing for a clean and structured overview of the reaction's details. |
| V2.0    | 09.01.2025 | Added addition order in AC2. |
| V3.0    | 10.02.2025 | Added ordering rules for input and outcome display. |

## 3. User Story

**AS A** user,  
**I WANT** to preview the reaction with an option to copy it as a .png image,  
**SO THAT** I can have a structured overview of the reaction's details for quick familiarization and use in presentations and documentation.

## 4. Acceptance Criteria

| User Role | AC# | GIVEN                              | WHEN                                          | THEN |
| --------- | --- | ---------------------------------- | --------------------------------------------- | ---- |
| User (All group members)      | 1   | The user is logged in to the system and has datasets with reactions to view | The user navigated and is viewing a dataset with a list of reactions or specific reaction page with details | The system displays a structured preview of the reaction always as a first top information about the reaction, where each input and outcome is clearly separated (Pic. 1). |
| User (All group members)     | 2   | The user is viewing the reaction preview | The user observes the inputs (all chemicals added to the reaction vessel) section | The system displays all reaction input components with their respective molecular structure drawings, names, amounts with units, roles in reaction, as well as addition order with respect to each Input. Each Input should be separated from the other. Additionally, all inputs should be demonstrated according to the addition order followed by alphabetical order, similarly as in [[ORD-6.6.2] Inputs tab view - inputs ordering](https://github.com/orgs/open-reaction-database/projects/7/views/11?pane=issue&itemId=95434044&issue=open-reaction-database%7Cord-app%7C237). |
| User  (All group members)    | 3   | The user is viewing the reaction preview | The user observes the reaction arrow that always clearly separates the inputs and outcomes | The system always displays the SHORT version of reaction conditions above the reaction arrow |
| User  (All group members)    | 4   | The user is viewing the reaction preview | The user observes the outcomes (products) section | The system shows all outcomes with their respective molecular structure drawings including top label (reaction time and the conversion % if available) and bottom label (yield % and reaction role if available under each chemical component),  both in SHORT format. Each Outcome should be separated from the other. Additionally, all Outcomes should be demonstrated according to the reaction time order if it's available, otherwise they should be displayed in the same order as they are stored. |
| User  (All group members)    |  5  | The user is viewing the reaction preview | The user observes the outcomes (products) section and wants to identify if the outcomes have the desired product | The system shows all outcomes and if the desired product is present that specific PRODUCT (inside Outcome) has a label of "desired product" (Pic. 2) |
| User  (All group members)    | 6  | The user is viewing the reaction preview | The molecular structure drawing for any reaction component is unavailable | The system shows instead of structural drawing the following information based on the first availability (the priority): compound's Name, or CAS number, or SMILES or any other available formats |
| User (All group members)     | 7   | The user is viewing the reaction preview (on reaction management page only - no need to have it in reaction preview on the dataset reaction list view) | The user clicks the "Copy Reaction Image" button (Pic. 3) | The system copies the reaction preview as a .png image to the clipboard, suitable for pasting into documents or presentations. Upon the successful completion of copy the system shows the notification: "Copied as PNG", otherwise it should show : "error in Copying as PNG" |
| User  (All group members)    | 8   | The user is viewing the reaction preview | The system shows the reaction that is very complex and long | The system has horizontal scrolls to let user see all the details |
| User  (Admin or Editor)      | 9  | The user is creating the new reaction or updating the existing one | The user is adding or editing all related information to Inputs, Conditions, and Products | The system automatically adds and updates that in the preview section |

<img width="1440" alt="Image" src="https://github.com/user-attachments/assets/0095b5fa-4ed0-4ba5-ac38-7d47eec7bcad" />
Pic. 1

<hr />

![Image](https://github.com/user-attachments/assets/120bc1ae-950b-4076-b872-235ee5944a4e)
Pic. 2

<hr />

<img width="1440" alt="Image" src="https://github.com/user-attachments/assets/b0ac49e0-ed75-4c1a-9e84-1e9004700a79" />
Pic. 3

## 5. Mockups Links
- [Dataset – List View](https://www.figma.com/design/c7M9fhVn278ei9rkepI2Sm/UX-UI-Open-Reaction-Database?node-id=1698-70946&t=PxW4QuHy1EOgu798-11)
- [Desired Product](https://www.figma.com/design/c7M9fhVn278ei9rkepI2Sm/UX-UI-Open-Reaction-Database?node-id=2080-17778&t=PxW4QuHy1EOgu798-11)
- [Reaction – Inputs – Filled](https://www.figma.com/design/c7M9fhVn278ei9rkepI2Sm/UX-UI-Open-Reaction-Database?node-id=3619-87303&t=PxW4QuHy1EOgu798-11)

## 6. User Interaction and Design

| # | Name | Type | Description |
| :-- | :-- | :-- | :-- |
| 1 | Reaction Preview | Display Element | Structured preview of a reaction displayed at the top as a generated image. |
| 2 | "Copy Reaction Image" Button | Button | A button that allows the user to copy the reaction preview as a PNG image to the clipboard. |
| 3 | Notification | Pop-up Message | A message displayed to the user to confirm the copying of the reaction preview as PNG or to show errors. |
| 4 | Horizontal Scroll | Navigation Control | Scroll control that appears when a reaction is too long or complex, allowing users to see all details. |
| 5 | Desired Product Label | Label/Tag | A label or tag that is visually associated with specific outcomes to indicate a 'desired product'. |

## 7. Non-functional Requirements **(OUT OF MVP SCOPE)**

- **Performance:** The reaction preview should load within 3 seconds to ensure a smooth user experience.
- **Scalability (NFR-4):** The system performance for preview generation and copying as .png should be planned for potential increases in user load up to 50 concurrent users without degradation.

## 8. Assumptions

- The system has the capability to render graphical representations of chemical structures from Ketcher.
- Users have a basic understanding of how to interpret chemical reaction diagrams.
- To perform any actions RBAC and Group Management (ORD-11.1; 11.4.2) must be considered for user.

## 9. Dependencies

- Dependent on ORD-6.5 Ketcher implementation
- The reaction preview feature is dependent on the integration of a graphical rendering tool capable of exporting images as .png files.

## 10. Out of Scope

| # | Description                               | User Story |
| - | ----------------------------------------- | ---------- |
|  | |  |

## 11. Questions

| # | Question                                              | Address to Whom       | Answer |
| - | ----------------------------------------------------- | --------------------- | ------ |
| 1 | What resolution and dimensions should the .png image be? | dev team / user | Will vary as it depends on the reactions and what browser will save |
| 2 | How short versions of conditions are now generated in reaction previews? | user | Backend function - parse the recording. Endpoint will be provided with information probably. |
| 3 | Outcome doesn't have name to be demonstrated? To demonstrate outcome name in preview Identifier NAME is needed to be added. Maybe Name Identifier should be provided by default for filling?  Currently 2 are shown when Ketcher is used. And small text near Identifiers can be added to explain what information will be shown in preview. | user | Names are not strongly required at the moment for both Inputs and Outcomes, since names are not very informative. |
| 4 | Confirm ALL details (for all reaction parts) that need to be demonstrated in the preview | user | Confirmed at Clarification sessions and on Slack |
| 5 | AC#5 - what priority to show compounds should be here? SMILES? and then what? | user | Molecular structure drawing IMAGE, then Name, then CAS number, then SMILES |
| 6 | What is expected to be shown in the Outcomes preview text? Currently most of the cases have time and yield. | user | Answer was provided in Slack |
| 7 | Another nice to have feature would be some highlight of the chemical component if it has the "desired product" = True field. | UI/UX | Special labeling or background color could be introduced for desired product. The decision on the exact way should be based on the best UI/UX practices. |
| 8 | 13.02.2025 It has been noted that the Reaction summary/preview on the Dataset page differs from the one on the Reaction page. Just to confirm that in our case we will have the same previews on Dataset page and Reaction page. Slack. | user | For new system - decided to have one version for both pages with all details. |

**Comment by ivanmaslow** [comment]
The feature has been successfully verified in the local environment:
- main branch, commit   e6b3d7adeb58ebc2dcd481cc2883d7c636af07ba;
- dev-z4acb31kcl4prqtw.us.auth0.com auth tenant;
- Desktop, Windows 11 OS, Chrome Browser.

Overview of How It Works:
- Preview on Reaction View is the same as on Dataset View, horizontal scroll is available:
![Image](https://github.com/user-attachments/assets/ab5d2588-59f4-454d-82c3-55b6e48dde47)
![Image](https://github.com/user-attachments/assets/6705d6b1-0d28-4f82-a426-a6fa5505f65f)
- No preview icon is displayed if there is no preview:
![Image](https://github.com/user-attachments/assets/410f945a-34e6-4128-8642-78143fe2716a)

Passed Tests:
- Verify that the reaction preview is displayed at the top, with inputs and outcomes clearly separated, on both the Dataset View and Reaction View pages
- Verify that the inputs section in the reaction preview displays molecular structures, names, amounts (with units), roles, and is ordered by addition order, then alphabetical order
- Verify that the desired product in the outcomes section is labeled as "desired product" if present
- Verify that when molecular structure drawings are unavailable, the system displays "No preview" icon
- Verify that clicking "Copy Reaction Image" button copies the reaction preview as a .png image to the clipboard, and displays notification: "Copied as PNG" or "Error in Copying as PNG
- Verify that horizontal scrolls are available for complex and long reactions, allowing the user to view all details in the reaction preview
- Verify that adding/editing/removing Inputs and Outcomes automatically updates the reaction preview section

Failed Tests:
- Verify that the reaction arrow clearly separates inputs and outcomes, and the SHORT version of reaction conditions is displayed above the arrow in the reaction preview	#580
- Verify that the outcomes section in the reaction preview displays molecular structures, reaction time, conversion, yield, and reaction role	#598
- Verify that the outcomes section is ordered by reaction time if available, or by stored order if reaction time is unavailable	#599
- Verify that clicking "Copy Reaction Image" allows pasting into a drawing tool with reaction labels fully visible and not truncated, matching the Preview	#587
- Verify that adding or editing Conditions automatically updates the reaction preview section	#580
- Verify that the display of measurement units on the reaction preview matches the format of the previous application	#436

Detected bugs are attached to the story and mentioned above.
